improve normalization of `Pointee::Metadata`

lukas-code opened the pull request:
This PR makes it so that `<Wrapper<Tail> as Pointee>::Metadata` is normalized to `<Tail as Pointee>::Metadata` if we don't know `Wrapper<Tail>: Sized`. With that, the trait solver can prove projection predicates like `<Wrapper<Tail> as Pointee>::Metadata == <Tail as Pointee>::Metadata`, which makes it possible to use the metadata APIs to cast between the tail and the wrapper:

```rust
#![feature(ptr_metadata)]

use std::ptr::{self, Pointee};

fn cast_same_meta<T: ?Sized, U: ?Sized>(ptr: *const T) -> *const U
where
    T: Pointee<Metadata = <U as Pointee>::Metadata>,
{
    let (thin, meta) = ptr.to_raw_parts();
    ptr::from_raw_parts(thin, meta)
}

struct Wrapper<T: ?Sized>(T);

fn cast_to_wrapper<T: ?Sized>(ptr: *const T) -> *const Wrapper<T> {
    cast_same_meta(ptr)
}
```

Previously, this failed to compile:

```
error[E0271]: type mismatch resolving `<Wrapper<T> as Pointee>::Metadata == <T as Pointee>::Metadata`
  --> src/lib.rs:16:5
   |
15 | fn cast_to_wrapper<T: ?Sized>(ptr: *const T) -> *const Wrapper<T> {
   |                    - found this type parameter
16 |     cast_same_meta(ptr)
   |     ^^^^^^^^^^^^^^ expected `Wrapper<T>`, found type parameter `T`
   |
   = note: expected associated type `<Wrapper<T> as Pointee>::Metadata`
              found associated type `<T as Pointee>::Metadata`
   = note: an associated type was expected, but a different one was found
```

(Yes, you can already do this with `as` casts. But using functions is so much :sparkles: *safer* :sparkles:, because you can't change the metadata on accident.)

---

This PR essentially changes the built-in impls of `Pointee` from this:

```rust
// before

impl Pointee for u8 {
    type Metadata = ();
}

impl Pointee for [u8] {
    type Metadata = usize;
}

// ...

impl Pointee for Wrapper<u8> {
    type Metadata = ();
}

impl Pointee for Wrapper<[u8]> {
    type Metadata = usize;
}

// ...

// This impl is only selected if `T` is a type parameter or unnormalizable projection or opaque type.
fallback impl<T: ?Sized> Pointee for Wrapper<T>
where
    Wrapper<T>: Sized
{
    type Metadata = ();
}

// This impl is only selected if `T` is a type parameter or unnormalizable projection or opaque type.
fallback impl<T /*: Sized */> Pointee for T {
    type Metadata = ();
}
```

to this:

```rust
// after

impl Pointee for u8 {
    type Metadata = ();
}

impl Pointee for [u8] {
    type Metadata = usize;
}

// ...

impl<T: ?Sized> Pointee for Wrapper<T> {
    // in the old solver this will instead project to the "deep" tail directly,
    // e.g. `Wrapper<Wrapper<T>>::Metadata = T::Metadata`
    type Metadata = <T as Pointee>::Metadata;
}

// ...

// This impl is only selected if `T` is a type parameter or unnormalizable projection or opaque type.
fallback impl<T /*: Sized */> Pointee for T {
    type Metadata = ();
}
```

compiler-errors commented:
r? types

lcnr commented on tests/ui/traits/pointee-normalize-equate.rs:
```diff
@@ -0,0 +1,55 @@
+// but `?Sized` considering regions.
+// Normalize `WrapperProject<&'a T>::Metadata` -> `<&'a T as Project>::Assoc::Metadata`
+// and don't require `WrapperProject<&'a T>: Sized`.
+fn sized_modulo_regions<'a, T: ?Sized + 'static>(
```
I think it's fine for this to result in an error. This matches the handling of where-bounds with other goals, e.g
```rust
trait Trait<'a> {}

impl<'a, T> Trait<'a> for T{}

fn impls_trait<'a, T: Trait<'a>>(_: &'a ()) {}

fn foo<'a, T: Trait<'static>>(x: &'a ()) {
    impls_trait::<T>(x)
}
```

lukas-code replied:
Alternatively, we could also always normalize from `Type::Metadata` to `Tail::Metadata` if `Type != Tail` and then make bounds like `Wrapper<T>: Sized` imply `T: Sized` (during elaboration?). That way all the cases in this test should still compile.

I will look into that,
@rustbot author

lukas-code replied:
Looks like making `Wrapper<T>: Sized` imply `T: Sized` is not straight forward, because it can break existing code by making trait selection ambiguous in a few edge cases involving higher-ranked bounds:

```rust
trait Project {
    type Assoc: ?Sized;
}

struct WrapperProject<T: ?Sized + Project>(T::Assoc);

// making (1) imply (2) causes error: "type annotations needed"
// because having both (2) and (3) in the param env makes
// `<&'a () as Project>::Assoc: Sized` ambiguous
fn foo<'a>()
where
    for<'b> &'b (): Project,
    for<'b> WrapperProject<&'b ()>: Sized, // (1)
    // for<'b> <&'b () as Project>::Assoc: Sized, // (2)
    <&'a () as Project>::Assoc: Sized, // (3)
{}
```

Because I don't know a clean solution for this problem at the moment, I've just made the `sized_modulo_regions` test case fail for now like you suggested.

lcnr commented on compiler/rustc_trait_selection/src/traits/project.rs:
```diff
@@ -2298,17 +2298,33 @@ fn confirm_builtin_candidate<'cx, 'tcx>(
+                );
+                let sized_obligation = obligation.with(tcx, sized_predicate);
+                if self_ty == tail
+                    || selcx.infcx.predicate_must_hold_considering_regions(&sized_obligation)
```
using `must_hold_considering_regions` outside of pure comp time optimizations should be avoided.

It results in weird changes between `Foo<'a>` and `Foo`. We should document this somewhere

lukas-code commented:
I replaced `predicate_must_hold_considering_regions` with `predicate_must_hold_modulo_regions` and added some docs to warn about `predicate_must_hold_considering_regions`.

r? @lcnr
@rustbot ready

lcnr commented on compiler/rustc_const_eval/src/interpret/terminator.rs:
```diff
@@ -372,14 +372,7 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
-                );
-                meta
-            };
+            let meta_ty = |ty: Ty<'tcx>| ty.ptr_metadata_ty(*self.tcx, |ty| ty);
```
(existing issue) why do we not normalize here? feels wrong. I expect that this can result in an ICE when using a type with an alias as a tail.

RalfJung replied:
Parts of the interpreter are called on generic code. So it'd be better to handle this with `TooGeneric`, like we already do elsewhere.

I just don't understand why we need to normalize here at all, since all types go through `subst_from_current_frame_and_normalize_erasing_regions` so they should already be normalized.

RalfJung replied:
In particular, we have a bunch of matches on `ty.kind()` above this line that seem like they may miss the types they check for if this can be non-normalized?

RalfJung replied:
I think the issue might be here:

https://github.com/rust-lang/rust/blob/16ba56c24238b2f44e0a97da760c83d8795abd80/compiler/rustc_const_eval/src/interpret/terminator.rs#L686-L692

We project to a field but then do not normalize again.

lukas-code replied:
> All `TyAndLayout` floating around in the interpreter must already be normalized. So adding normalization here is _not_ a good idea, it just masks an issue somewhere else where we are failing to normalize.

Even if all types are normalized, a field of a normalized ADT can still have an unnormalized type.

For example here

```rust
trait Project {
    type Assoc;
}

struct Foo<T: Project> {
    x: T::Assoc,
}
```

`Foo<()>` is a normalized type, but the field `x` of `Foo` has type `<() as Project>::Assoc`. To find the metadata of an ADT, we need to look at its tail and normalize these projections away. That's what the `normalize` function passed as the second parameter of `ptr_metadata_ty` is for. If the normalize function returns an unnormalized type,  then `ptr_metadata_ty` will ICE after this PR or return false for the second output before this PR.

> I think the issue might be here:
> 
> https://github.com/rust-lang/rust/blob/16ba56c24238b2f44e0a97da760c83d8795abd80/compiler/rustc_const_eval/src/interpret/terminator.rs#L686-L692
>
> We project to a field but then do not normalize again.

The types of tuple fields are always fully normalized if the tuple type itself is fully normalized. You can't have a tuple `(..., T, ...)` and the field be `<T as Project>::Assoc` or something like you can with ADTs.

RalfJung replied:
> The types of tuple fields are always fully normalized if the tuple type itself is fully normalized. 

Ah okay. We should probably have a comment there to document why we don't need to normalize.

> Even if all types are normalized, a field of a normalized ADT can still have an unnormalized type.

Okay. But that still doesn't mean that this is the right place to normalize. We should normalize wherever the unnormalized type creeps into the interpreter -- probably in the `project_field` function.

RalfJung replied:
I don't see how that reply related to what I said.

The bug is that someone is passing non-normalized types around the interpreter. Various parts of the interpreter assume things are normalized. The code in `layout_compat` is correct, some other code is wrong for hading non-normalized types to `layout_compat`.

RalfJung replied:
Oh you are saying even if we have a normalized type in `layout_compat` it can still hit non-normalized types further down. Now I see. This needs a big fat warning in the `ptr_metadata_ty` docs.

Nevertheless is still also seems wrong that `project_field` does not normalize? Or is that fine?

lukas-code replied:
We could remove the normalize parameter from `ptr_metadata_ty` entitely and always use `normalize_erasing_regions` with `ParamEnv::reveal_all()`. Or can there be any type information in the `InterpCx`'s param env that is required for this normalization?

lcnr replied:
it's also required by typeck/borrowck where normalization must not use `normalize_erasing_regions` as we need to handle region constraints from normalization

lukas-code replied:
I moved the miri changes to https://github.com/rust-lang/rust/pull/120587.

lcnr commented on compiler/rustc_trait_selection/src/traits/project.rs:
```diff
@@ -1919,7 +1919,7 @@ fn assemble_candidates_from_impls<'cx, 'tcx>(
                         // metadata if they're known (e.g. by the param_env) to be sized
                         ty::Param(_) | ty::Alias(..)
-                            if selcx.infcx.predicate_must_hold_modulo_regions(
+                            if self_ty != tail || selcx.infcx.predicate_must_hold_modulo_regions(
```
why does this only apply if `self_ty != tail`? :thinking:

lcnr commented on tests/ui/traits/pointee-normalize-equate.rs:
```diff
@@ -0,0 +1,41 @@
+// check-pass
+
```
```suggestion
// revisions: next old
//[next] compile-flags: -Znext-solver
```
I can help if the next solver does not already support this

lukas-code replied:
The new solver does indeed not support this:
```
error[E0271]: type mismatch resolving `<T as Pointee>::Metadata == <Wrapper<T> as Pointee>::Metadata`
  --> src/lib.rs:21:20
   |
21 |     cast_same_meta(ptr)
   |     -------------- ^^^ types differ
   |     |
   |     required by a bound introduced by this call
   |
```

The relevant code seems to be here and it looks like it's already supposed to normalize from `Wrapper<T>::Metadata` to `T::Metadata`:

https://github.com/rust-lang/rust/blob/af08c64e3800fd15f0afdb8651de8fa7dacc0026/compiler/rustc_trait_selection/src/solve/normalizes_to/mod.rs#L410-L435

I don't know anything about the new solver, so any help to get this working would be appreciated.

lukas-code replied:
Update: I've pushed a commit that appears to make it work with the new solver, but I don't know what I'm doing so please review carefully.

lcnr commented on compiler/rustc_trait_selection/src/solve/normalizes_to/mod.rs:
```diff
@@ -379,6 +379,45 @@ impl<'tcx> assembly::GoalKind<'tcx> for NormalizesTo<'tcx> {
+                        ecx.evaluate_added_goals_and_make_canonical_response(Certainty::Yes)
+                    });
+
+                if let Ok(response) = sized_probe {
+                    return Ok(response);
+                }
```
actually, this isn't as trivial as I originally thought. Builtin traits should generally follow the same rules as non-builtin traits. If they don't, it's very easy to accidentally break an invariant of the type system.

Right now you pretty much add these builtin impls
```rust
struct MyType<T: ?Sized> {
    last: Tail<T>,
}

impl<T: ?Sized> Pointee for MyType<T>
where
    Self: Sized,
{
    type Metadata = ();
}

impl<T: ?Sized> Pointee for MyType<T> {
    type Metadata = <Tail<T> as Pointee>::Assoc;
}
```
These impls overlap. While I can't think of concrete problem caused by this overlap, it should still be avoided if possible. We do not allow overlapping impls with associated types anywhere else (apart from the "dyn Trait: Trait" impl, which is currently causing an actual soundness bug #57893). Now, the associated type of these two impls should always end up being the same, so it probably does not cause an unsoundness that way. But having impls like this can also result in ambiguity. Regardless, it complicates the reasoning about the type system.

Thinking about this, we should be able to only emit the second impl, i.e.
```rust
impl<T: ?Sized> Pointee for Adt {
    type Metadata = <Tail<T> as Pointee>::Assoc;
}
```
Because types cannot be recursive without any indirection, this should bottom out in a type for which we can trivially compute the metadata.

lcnr replied:
> Is this expected behavior? With manual impls you get an ambiguity error:

no, this is because of your implementation. This way of handling multiple candidates (by only considering the second one if the first one does not apply) is something we avoid elsewhere. We instead collect all candidates and then use `try_merge_candidates` to actually get to a single one.

> You could argue that's fine to break, because `Pointee` is unstable and nobody writes code like that anyway. However, breaking this code would create situations where it's possible to cast a thin pointer to `*const Wrapper<Tail>`, but not know `Wrapper<Tail>: Thin`. The entire purpose of this PR is bringing the checks of pointer-to-pointer casts and the builtin impl of the `Pointee` trait closer together

I agree that it is weird to be able to pointer cast while not actually being able to prove that the cast is correct in the type system. However, I would still really like to avoid these overlapping impls. Can pointer casting be changed to rely on `Metadata` being equal instead of using a separately written check?

lukas-code replied:
> no, this is because of your implementation.

This compiles on nightly, without my implementation.

```rust
#![feature(ptr_metadata)]
use std::ptr::Pointee;

fn check_meta<T, U: Pointee<Metadata = T>>() {}

fn sized_slice<T>()
where
    [T]: Sized,
{
    // We have `[T]: Sized`, but the metadata is still `usize`.
    check_meta::<usize, [T]>(); // compiles
}
```

What I'm trying to say is that if my implementation is wrong, we should change this to be an ambiguity error as well.

---

> Can pointer casting be changed to rely on `Metadata` being equal instead of using a separately written check?

Checking for equal metadata alone is not enough, because that can't cover fat->thin casts or casts between pointers to trait objects that change the trait object lifetime or add auto traits. We'd need something like a new built-in trait `MetadataCompatible<T>` (or new predicate kind) to fully typecheck pointer-to-pointer casts with traits. I've actually been working on a [proposal](https://hackmd.io/BcsyukxVSm-mLPY4Rlj5mg) and [experimental implementation](https://github.com/lukas-code/rust/pull/6) to move pointer-to-pointer casts to the trait system entirely. This PR is the first step to making that happen and I was hoping it would be sufficiently motivated by just the `cast_same_meta` example from the OP and "let's not break code that currently compiles", but evidently that's not the case, so here we are.

---

Would it perhaps be possible to treat the current implementation as a specialization, i.e.

```rust
impl<T: ?Sized> Pointee for MyType<T>
where
    T: Pointee,
{
    default type Metadata = <T as Pointee>::Metadata;
}

impl<T: ?Sized> Pointee for MyType<T>
where
    T: Pointee,
    T: Sized,
{
    type Metadata = ();
}
```

That appears to compile on nightly, but idk if it's actually sound.

Otherwise, I guess just normalizing to the metadata of the tail is an improvement over the status quo and all cases that it can break are easily fixed by adding `where Tail: Sized`.

lukas-code replied:
(clarifying previous comments)

The current impls for `Pointee` on nightly looks something like this:

```rust
impl Pointee for u8 {
    type Metadata = ();
}

impl Pointee for [u8] {
    type Metadata = usize;
}

impl Pointee for Wrapper<u8> {
    type Metadata = ();
}

impl Pointee for Wrapper<[u8]> {
    type Metadata = usize;
}

// this impl is only selected if no other candidate applies
fallback impl<T: Sized> Pointee for T {
    type Metadata = ();
}
```

That means that if we know `T == [u8]` and `T: Sized`, then `T::Metadata == usize`, but if we only know `T: Sized`, then `T::Metadata == ()`.

The impls in this PR as is will extend the above by another fallback case:
```rust
impl Pointee for u8 {
    type Metadata = ();
}

impl Pointee for [u8] {
    type Metadata = usize;
}

impl Pointee for Wrapper<u8> {
    type Metadata = ();
}

impl Pointee for Wrapper<[u8]> {
    type Metadata = usize;
}

// this impl is only selected if no candidate above applies
fallback impl<T: Sized> Pointee for T {
    type Metadata = ();
}

// this impl is only selected if no candidate above applies, including the other fallback
fallback fallback impl<T: ?Sized> Pointee for Wrapper<T> {
    type Metadata = <T as Pointee>::Metadata;
}
```

An alternative approach that doesn't rely on such fallbacks, which can't currently be expressed in the type system, would instead rely on specialization:

```rust
impl Pointee for u8 {
    default type Metadata = ();
}

impl Pointee for [u8] {
    default type Metadata = usize;
}

impl<T: ?Sized> Pointee for Wrapper<T> {
    default type Metadata = <T as Pointee>::Metadata;
}

// this impl is preferred over all other impls
specialize impl<T: Sized> Pointee for T {
    type Metadata = ();
}
```

With that, if we know `T == [u8]` and `T: Sized`, then `T::Metadata == ()`, because the specialization will take priority over the defaults.

RalfJung commented on compiler/rustc_const_eval/src/interpret/terminator.rs:
```diff
@@ -372,14 +372,8 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
-                );
-                meta
-            };
+            let normalize = |ty: Ty<'tcx>| self.tcx.normalize_erasing_regions(self.param_env, ty);
```
```suggestion
            // Even if `ty` is normalized, the search for the unsized tail will project to fields
            // which can yield non-normalized types. So we need to provide a normalization
            // function.
            let normalize = |ty: Ty<'tcx>| self.tcx.normalize_erasing_regions(self.param_env, ty);
```

lukas-code replied:
I will extract the normalization change to another PR.

bors commented:
:umbrella: The latest upstream changes (presumably #120660) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

lcnr commented:
> I was hoping that this PR would be sufficiently motivated by just the `cast_same_meta` example from the OP and "let's not break code that currently compiles", but evidently that's not the case, so here we are.

I agree that `cast_same_meta` should work, not sold on "let's not break code that currently compiles". I would expect that also using the adt tail for pointer casts won't break any actual code, even though it is theoretically breaking. I would very much prefer to try and break this for a "cleaner language" and only think about how to handle these overlapping impls if too many users are effected. We have `crater` for this.

I'd therefore be in favor of "Otherwise, I guess just normalizing to the metadata of the tail is an improvement over the status quo and all cases that it can break are easily fixed by adding `where Tail: Sized`."

lukas-code commented:
@rustbot author

Commit:
extend docs for `predicate_must_hold_considering_regions`

Commit:
old solver: improve normalization of `Pointee::Metadata`

Commit:
new solver: improve normalization of `Pointee::Metadata`

lcnr commented on compiler/rustc_trait_selection/src/solve/normalizes_to/mod.rs:
```diff
@@ -422,13 +424,11 @@ impl<'tcx> assembly::GoalKind<'tcx> for NormalizesTo<'tcx> {
-                        let self_ty = field_def.ty(tcx, args);
+                    Some(tail_def) => {
+                        let tail_ty = tail_def.ty(tcx, args);
+                        let predicate = goal.predicate.with_self_ty(tcx, tail_ty).as_projection();
```
instead of adding it as a nested goal, simply return the `<tail_ty as Pointee>::Metadata`. If the caller needs a fully normalized type, they call `NormalizesTo` in a loop.

I think you're using `as_projection()` here as otherwise a tail whose metadata should remain rigid would result in a n error. However, `NormalizesTo` in the new solver has to only normalize by 1 step. This differs from the old solver where `project` has to deeply normalize.

lukas-code commented:
I've updated the OP to show the differences between before and after this PR. I'm still not convinced that the "fallback impl" is better than a specialization, but I guess it doesn't matter too much. :shrug:

@rustbot ready

lcnr commented:
> I've updated the OP to show the differences between before and after this PR. I'm still not convinced that the "fallback impl" is better than a specialization, but I guess it doesn't matter too much. :shrug:

Neither am I. We already had that impl where we check for `Sized` for aliases and params. It is pretty much just an `AliasBound` candidate instead of an actual impl.

Pretty much what happens here is that if there's a `Sized` bound in the environment, either directly via `T: Sized` or indirectly from an item bound of `T: Trait` with `type Assoc: Sized`, we assume this means `Pointee<Metadata = ()>`. This is something the caller should prove. The only way you have an alias or an opaque as the self type (at least in the new solver) is if they are rigid as normalizing them relies on *the environment*.

We hackily reimplemented a super trait bound: `trait Sized: Pointee<Metadata = ()> {}`. I would like us to remove this manual builtin impl and instead add that super trait bound. That doesn't have to happen inside of this PR. I feel like we may have even considered to do so in the past but wasn't able to find anything skimming through previous PRs.

@lukas-code would you like to change it to a super trait right away or should we leave it to a future PR? If you want to not do so rn, please add a `FIXME` for it to the "fallback impl" arm in both solvers

lukas-code commented:
Making `Pointee<Metadata = ()>` a supertrait of `Sized` seems like a really good idea. I'll implement that and see what breaks 😆 .

Commit:
add fixme

lukas-code commented:
To answer the question of what breaks, it's this test: (in both solvers) https://github.com/rust-lang/rust/blob/4a2fe4491ea616983a0cf0cbbd145a39768f4e7a/tests/ui/sized/recursive-type-pass.rs#L1-L10

But that's probably fine, because that test compiling at all seems more like a side-effect of `Sized` being coinductive rather than intentional behavior. All other tests that rely on `Sized` being coinductive still pass

But there are also a ton of diagnostics regressions that now complain about "type mismatch resolving `<Foo as Pointee>::Metadata == ()`" in addition to "`Foo` is not sized" and fixing those seems to be non-trivial, so I'd like to move making `Pointee<Metadata = ()>` a supertrait of `Sized` to a future PR.

@rustbot ready

lukas-code commented on compiler/rustc_trait_selection/src/solve/normalizes_to/mod.rs:
```diff
@@ -410,7 +410,10 @@ impl<'tcx> assembly::GoalKind<'tcx> for NormalizesTo<'tcx> {
                 }
 
                 ty::Alias(_, _) | ty::Param(_) | ty::Placeholder(..) => {
-                    // FIXME(ptr_metadata): It would also be possible to return a `Ok(Ambig)` with no constraints.
```
This would need to return `Err(NoSolution)`, not `Ok(Ambig)`.

lcnr commented:
thanks for your good work and for staying with me while I tried to figure this out myself :heart: 

@bors r+
